graphql_compiler/compiler/emit_sql.py

```diff
@@ -239,19 +340,152 @@ def start_global_operations(self):
 
     def filter(self, predicate):
         """Execute a Filter Block."""
-        sql_expression = predicate.to_sql(self._aliases, self._current_alias)
+        if self._in_fold or isinstance(predicate.left, FoldedContextField) \
+                or isinstance(predicate.right, FoldedContextField):
+            raise NotImplementedError('Filters inside a fold are not implemented yet.')
+
+        sql_expression = predicate.to_sql(self._aliases,
+                                          self._current_alias,
+                                          self._fold_output_column_aliases)
         if self._is_in_optional_scope():
             sql_expression = sqlalchemy.or_(sql_expression,
                                             self._came_from[self._current_alias].is_(None))
         self._filters.append(sql_expression)
 
+    def fold(self, fold_scope_location):
+        """Begin execution of a Fold Block."""
+        # Subquery will look as follows:
+        #
+        # SELECT
+        #   OuterVertex.uuid <- this value is determined from the edge descriptor
+        #   array_agg(InnerVertex.fold_output_column) AS fold_output
+        # FROM OuterVertex
+        # INNER JOIN ... <- INNER JOINs compiled during Unfold block
+        # ON ...
+        #          ...
+        # INNER JOIN InnerVertex <- INNER JOINs compiled during Unfold block
+        # ON ...
+        # GROUP BY OuterVertex.uuid
+
+        if self._in_fold:
+            raise AssertionError(u'Reached an invalid fold state.')
+
+        # begin the fold
+        self._in_fold = True
+        # outer vertex is whatever vertex occurs immediately before the fold
+        self._fold_outer_vertex_location = self._current_location
+        self._fold_inner_vertex_location = fold_scope_location
+
```

Comment:
Shouldn't  `self._current_location` be updated to be the `fold_scope_location`? That's what the spec for that field implies should happen.

Reply:
I don't think so because we don't actually traverse into the folded location. If we did, then I think we'd need to backtrack after the Unfold.

Imagine we have 
```
{
        Animal {
            name @output(out_name: "animal_name")
            out_Animal_ParentOf @fold {
                name @output(out_name: "child_names_list")
            }
            in_Animal_ParentOf @optional {
                name @output(out_name: "parent_name")
            }
        }
    }
```

If we changed current location inside the Fold block, there would be no relocate before the Traverse of in_Animal_ParentOf. Basically, I don't think the IR treats folds as changes of location like it does a regular traversal, because there is no backtrack or traverse call.